clients/src/main/java/org/apache/kafka/clients/consumer/internals/MetadataFetcher.java

```diff
@@ -0,0 +1,805 @@
+                        else if (error == Errors.UNKNOWN_TOPIC_OR_PARTITION)
+                            // if a requested topic is unknown, we just continue and let it be absent
+                            // in the returned map
+                            continue;
+                        else if (error.exception() instanceof RetriableException)
+                            shouldRetry = true;
+                        else
+                            throw new KafkaException("Unexpected error fetching metadata for topic " + topic,
+                                    error.exception());
+                    }
+                }
+
+                if (!shouldRetry) {
+                    HashMap<String, List<PartitionInfo>> topicsPartitionInfos = new HashMap<>();
+                    for (String topic : cluster.topics())
+                        topicsPartitionInfos.put(topic, cluster.partitionsForTopic(topic));
+                    return topicsPartitionInfos;
+                }
+            }
+
+            timer.sleep(retryBackoffMs);
+        } while (timer.notExpired());
+
+        throw new TimeoutException("Timeout expired while fetching topic metadata");
+    }
+
+    /**
+     * Send Metadata Request to least loaded node in Kafka cluster asynchronously
+     * @return A future that indicates result of sent metadata request
+     */
+    private RequestFuture<ClientResponse> sendMetadataRequest(MetadataRequest.Builder request) {
+        final Node node = client.leastLoadedNode();
+        if (node == null)
+            return RequestFuture.noBrokersAvailable();
+        else
+            return client.send(node, request);
+    }
+
+    /**
+     * Reset offsets for all assigned partitions that require it.
+     *
+     * @throws org.apache.kafka.clients.consumer.NoOffsetForPartitionException If no offset reset strategy is defined
+     *   and one or more partitions aren't awaiting a seekToBeginning() or seekToEnd().
+     */
+    public void resetOffsetsIfNeeded() {
+        // Raise exception from previous offset fetch if there is one
+        RuntimeException exception = cachedListOffsetsException.getAndSet(null);
+        if (exception != null)
+            throw exception;
+
+        Set<TopicPartition> partitions = subscriptions.partitionsNeedingReset(time.milliseconds());
+        if (partitions.isEmpty())
+            return;
+
+        final Map<TopicPartition, Long> offsetResetTimestamps = new HashMap<>();
+        for (final TopicPartition partition : partitions) {
+            Long timestamp = offsetResetStrategyTimestamp(partition);
+            if (timestamp != null)
+                offsetResetTimestamps.put(partition, timestamp);
+        }
+
+        resetOffsetsAsync(offsetResetTimestamps);
+    }
+
+    /**
+     * Validate offsets for all assigned partitions for which a leader change has been detected.
+     */
+    public void validateOffsetsIfNeeded() {
```

Comment:
I'd recommend renaming all functions like `validateOffsetsIfNeeded` / `resetOffsetsIfNeeded` here by replacing `Offset` to `Position`, since 1) we do not only just validate / reset offset value, but the whole position including epoch etc as well, 2) this is also to distinguish from the other `offsets` like the ones from fetch responses as well moving on.

Reply:
I had assumed that offsets and positions were synonyms. Is that incorrect?

I'll rename the methods as suggested.

Reply:
They used to be the same, but then later as we add more fields to Position it now becomes

```
class FetchPosition {
        public final long offset;
        final Optional<Integer> offsetEpoch;
        final Metadata.LeaderAndEpoch currentLeader;
}
```

And when we reset/validate, we'd touch on other fields than the offsets itself.

Another rationale is to distinguish the `position offset` from the `fetch offset` :)